File: Year-2023/LeetCode-February-Challenge/PermutationinString.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int s1Size = s1.size();
        int s2Size = s2.size();

        if(s1Size > s2Size) return false;

        vector<int> s1Freq(26, 0);
        vector<int> window(26, 0);

        for(int i = 0; i < s1Size; ++i){
            s1Freq[s1[i] - 'a']++;
            window[s2[i] - 'a']++;
        }
        if(window == s1Freq) return true;

        for(int i = s1Size; i < s2Size; ++i){
            window[s2[i - s1Size] - 'a']--;
            window[s2[i] - 'a']++;
            if(window == s1Freq) return true;
        }
        return false;
    }
};
int main(){
    string s1 = "ab", s2 = "eidbaooo";
    Solution *s = new Solution();
    cout<<std::boolalpha<<s->checkInclusion(s1, s2)<<endl;
    return 0;
}

File: Year-2023/LeetCode-February-Challenge/PermutationinString.js
var checkInclusion = function(s1, s2) {
    let s1Size = s1.length;
    let S2Size = s2.length;
    if(s1Size > S2Size) return false;
    let window = new Array(26).fill(0);

    for(let i = 0; i < s1Size; ++i){
        window[s1.charCodeAt(i) - 97]++;
        window[s2.charCodeAt(i) - 97]--;
    }
    if(!window.some(e => e !== 0)) return true;

    for(let i = s1Size; i < S2Size; ++i){
        window[s2.charCodeAt(i) - 97]--;
        window[s2.charCodeAt(i - s1Size) - 97]++;
        if(!window.some(e => e !== 0)) return true;
    }
    return false;
};

let s1 = "ab", s2 = "eidbaooo";
console.log(checkInclusion(s1, s2));